lib/bap_disasm/bap_disasm_rec.ml

```diff
@@ -213,26 +203,75 @@ let next dis s =
 
 let stop_on = [`Valid]
 
-(* find intersections, i.e. instructions which share bytes, but have different address,
-   and update destinations for those of them, which also share max address to
-   split a block in such place *)
+module Dest = struct
+  let rec compare_edge e e' =
+    if e = e' then 0
+    else
+      match e, e' with
+      | `Fall, _ -> -1
+      | `Jump, _ -> 1
+      | _ -> - (compare_edge e' e)
+
+  let compare (d,e) (d',e') =
+    match d, d' with
+    | None,None -> compare_edge e e'
+    | Some _, None -> 1
+    | None, Some _ -> -1
+    | Some d, Some d' ->
+       let x = Word.compare d d' in
+       if x = 0 then compare_edge e e'
+       else x
+
+  module Set = Set.Make(struct
```

Comment:
Do not create Set modules in the new code. Derive comparable, and use polymorphic sets, e.g.,

```ocaml
module Dest = struct 
   type t = addr option * edge [@@deriving compare, sexp]
   include Base.Comparable.Make(struct type nonrec t = t end)
end
...
let dests = Set.empty (module Dests)
```